AIP-86 - Remove no-longer-needed deadlines when a dagrun succeeds

ferruzzi opened the pull request:
When a DagRun finishes successfully, if that run has any Deadlines which have not expired, then we can safely remove them from the database.  If the Deadline has passed then it will be retained so we can still act on it in the next Scheduler pass.

This follows the previously-discussed policy that successful (not-needed) deadlines can be removed, and those which were missed will be retained in a separate table, which will allow UI integration and historical data on deadline miss statistics.  There will be a follow-up from @ramitkataria to handle missed deadlines while I start working on the next step to have the scheduler make the decision if there are deadlines which need to be acted on.


<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->



<!-- Please keep an empty line above the dashes. -->
---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [airflow-core/newsfragments](https://github.com/apache/airflow/tree/main/airflow-core/newsfragments).

ferruzzi committed:
AIP-86 - Remove no-longer-needed deadlines when a dagrun succeeds

When a DagRun finished successfully, if that run has any Deadlines which have not expired, then we can safely remove them from the database.  If the Deadline has passed then it will be retained so we can still act on it in the next Scheduler pass.

ferruzzi commented on airflow-core/src/airflow/models/deadline.py:
```diff
@@ -98,6 +98,61 @@ def _determine_resource() -> tuple[str, str]:
+        NOTE: This should only be used to remove deadlines which are associated with
+            successful dagruns. If the deadline was missed, move it to the `missed_deadlines`
+            table after executing the callback.
+        TODO:  Create the missed_deadlines table (Ramit)
```
@ramitkataria - Please remove this todo when you get this part done.

ferruzzi commented on airflow-core/src/airflow/models/deadline.py:
```diff
@@ -98,6 +98,61 @@ def _determine_resource() -> tuple[str, str]:
+
+        logger.debug("%d deadline records were deleted matching the conditions %s", deleted_count, conditions)
+
+        # Refresh any affected DAG runs.
+        for dagrun in dagruns_to_refresh:
+            session.refresh(dagrun)
```
I debated this part.  I could theoretically do this in the `for deadline, dagrun...` loop above, but a dagrun may have multiple deadlines so this is theoretically more efficient.  I can move it into the loop above if anyone feels strongly about it.

vincbeck replied:
I agree with this approach, I like it better

ramitkataria replied:
Me too!

ferruzzi commented on airflow-core/tests/unit/models/test_dagrun.py:
```diff
@@ -1256,13 +1256,14 @@ def test_dagrun_success_deadline(self, dag_maker, session):
             on_success_callback=on_success_callable,
             deadline=DeadlineAlert(
-                reference=DeadlineReference.FIXED_DATETIME(DEFAULT_DATE),
+                reference=DeadlineReference.FIXED_DATETIME(future_date),
```
This was working with a date in the past because we weren't yet checking if the deadline had passed.  Now that we are checking it, we need to use a date int he future here.

ramitkataria commented on airflow-core/src/airflow/models/deadline.py:
```diff
@@ -98,6 +98,61 @@ def _determine_resource() -> tuple[str, str]:
+            deadline_dagrun_pairs = (
+                session.query(Deadline, DagRun).join(DagRun).filter(and_(*filter_conditions)).all()
+            )
+        except SQLAlchemyError as e:
```
Should we use a subclass of `SQLAlchemyError` to narrow down types of exceptions being caught here?

ferruzzi replied:
I'll do a little more manual testing with bad column names and double check the exact error and message, see if I can narrow that down any.

ferruzzi replied:
I put that in there for some edge case and can't figure out what ti was.   In the end, an invalid column will raise an AttributeError on the model before it ever reaches a SQLAlchemy error, so I'm just going to scope that down to AttributeError, log it, rollback the change, and re-raise.  Running static checks now and I'll get that fix up in the morning.

ramitkataria commented on airflow-core/src/airflow/models/deadline.py:
```diff
@@ -98,6 +98,61 @@ def _determine_resource() -> tuple[str, str]:
+                session.query(Deadline, DagRun).join(DagRun).filter(and_(*filter_conditions)).all()
+            )
+        except SQLAlchemyError as e:
+            invalid_column = next(iter(conditions.keys()))  # Get the first key that caused the error.
```
I'm wondering how this line finds the key that caused the error 🤔

ramitkataria reviewed (approved):
Left 1 nit and a question but looks good otherwise!

ferruzzi committed:
PR fixes

- Renamed Deadline.remove_deadlines to prune_deadlines to indicate that the removal is conditional.
- Adjusted some comments to try to clarify that.
- Fix exception handling in the prune_deadline method.
- Remove the unit test that is no longer needed from the above changes.

o-nikolas commented on airflow-core/src/airflow/models/deadline.py:
```diff
@@ -98,6 +98,61 @@ def _determine_resource() -> tuple[str, str]:
+            invalid_column = next(iter(conditions.keys()))  # Get the first key that caused the error.
+            message = f"Invalid column '{invalid_column}' specified in conditions while resolving deadlines. Rolling back changes."
+            logger.exception(message)
+            session.rollback()
```
What's being rolled back here? Wasn't there just a query not a write/transaction?

ferruzzi replied:
I think you are right, I'll double check, but this may be unnecessary since the session.scalar should be a read-only transaction and the error would actually be thrown before the db is actually hit (while assembling the query).

o-nikolas commented on airflow-core/src/airflow/models/deadline.py:
```diff
@@ -98,6 +98,61 @@ def _determine_resource() -> tuple[str, str]:
+
+        NOTE: This should only be used to remove deadlines which are associated with
+            successful dagruns. If the deadline was missed, move it to the `missed_deadlines`
+            table after executing the callback.
```
Is the move intended to happen here? I wonder if we should rename this thing to something like `process_deadlines` since it feels like a bit more than deletion might happen here.

ferruzzi replied:
No, I think the move would be handled in one place where we handle the callbacks.

And on that note, @ramitkataria did some benchmarking and it looks like we may not be moving after all.  He'll follow up with that but the time to search an index vs a separate table is negligible even at a pretty huge scale.  I may have just been over-engineering it out of an abundance of caution.  More to come on that later.

o-nikolas replied:
Yeah synced with Ramit about the table vs column, that all sounds good. If that happens elsewhere I guess this function name is good enough. Although, I will note that it is still quite conditional in some cases not all, or even any, deadlines will be deleted. So it's more like a prune (which is a name you suggested when we chatted in Slack and I like that).

ferruzzi replied:
Yeah, changed to prune in the last revision, going to resolve this one.  It's a much better name and implies that we may or may not be actually deleting, depending on the condition.

o-nikolas commented on airflow-core/src/airflow/models/dagrun.py:
```diff
@@ -1217,6 +1218,10 @@ def recalculate(self) -> _UnfinishedStates:
 
+            if (deadline := dag.deadline) and isinstance(deadline.reference, DeadlineReference.TYPES.DAGRUN):
+                # The dagrun has succeeded, so the deadline is no longer needed.
+                Deadline.remove_deadlines(session=session, conditions={DagRun.run_id: self.run_id})
```
Why are `conditions` being exposed here? Seems just a tiny bit strange. Shouldn't this just take a param `run_id=self.run_id` and the conditions can be built within the function? Not sure why that's leaking out to the caller.

ferruzzi replied:
What happens when the next deadline type needs Asset.created_at or Job.start_date or whatever else we add in the future?  Wrapping this in a kwargs-like dict felt like the most reasonable solution.

o-nikolas replied:
Yeah that's fair. You could have different methods for each use case which assemble the correct conditions internally and then call some shared code for the actually processing of the conditions (basically what you have now). But, very fair if you want to push back on that. Not much internal logic is leaking here so it's somewhat of an academic nit.

ferruzzi replied:
Yeah, I had considered using kwargs and filtering them out but it loked and felt better explicitly using a parameter, and I feel the name itself fits, too. The method call reads pretty straight forward as "prune deadlines using this session which meet these conditions".  If you're cool with it, I'll resolve this one.

o-nikolas replied:
Yupp all good :+1:

o-nikolas commented on airflow-core/src/airflow/models/dagrun.py:
```diff
@@ -1217,6 +1218,10 @@ def recalculate(self) -> _UnfinishedStates:
                 )
 
+            if (deadline := dag.deadline) and isinstance(deadline.reference, DeadlineReference.TYPES.DAGRUN):
+                # The dagrun has succeeded, so the deadline is no longer needed.
```
```suggestion
                # The dagrun has succeeded, so the deadline(s) may no longer be needed.
```

Or just completely refactor.

o-nikolas commented on airflow-core/tests/unit/models/test_deadline.py:
```diff
@@ -103,6 +105,50 @@ def test_add_deadline(self, dagrun, session):
+        ],
+    )
+    @mock.patch("sqlalchemy.orm.Session")
+    def test_resolve_deadlines(self, mock_session, conditions):
```
```suggestion
    def test_remove_deadlines(self, mock_session, conditions):
```

The code under test seems to be the remove right? Same for below

ferruzzi replied:
`test_resolve_deadlines_invalid_column` for sure is no longer needed once I change that exception block as discussed above.   I think `test_resolve_deadlines` should still be valid though, no?

ferruzzi replied:
Ah, nevermind, I misread your comment, you were suggesting a rename of the test, not that the code was removed :sweat_smile:    Yeah, I can rename that.  The other one is still getting dropped though

o-nikolas replied:
Yeah, rename was the suggestion for sure! I should have been more clear :pray:
